### PR TITLE
Secret store (part2 - encrypted key/value svc)

### PR DESCRIPTION
**Pull request by NikVolf**
- SecureStore implements EncryptedHashMap for storing arbitrary data in key-files according ethereum client specs

### CODE REVIEW

**Comment by NikVolf** [comment]
coverage is kind of misleading here
`keys.rs` coverage hasn't changed
and `store.rs` is close to 100%

File: util/src/keys/directory.rs [hunk 1]
```diff
@@ -476,29 +502,66 @@ impl KeyDirectory {
...
+		for entry in try!(fs::read_dir(&self.path)) {
+			let entry = try!(entry);
+			if !try!(fs::metadata(entry.path())).is_dir() {
+				match entry.file_name().to_str() {
+					Some(ref name) => {
+						if let Ok(uuid) = uuid_from_string(name) { result.push(uuid); }
+					},
+					None => { continue; }
+				};
+
```
Comment (gavofyork) [inline util/src/keys/directory.rs]: superfluous newline

File: util/src/keys/directory.rs [hunk 2]
```diff
@@ -476,29 +502,66 @@ impl KeyDirectory {
...
-			self.cache_usage.drain(..untracked_usages);
+			cache_usage.drain(..untracked_usages);
 		}
 
-		if self.cache.len() <= MAX_CACHE_USAGE_TRACK { return; }
+		let mut cache = self.cache.borrow_mut();
+		if cache.len() <= MAX_CACHE_USAGE_TRACK { return; }
 
-		let uniqs: HashSet<&Uuid> = self.cache_usage.iter().collect();
+		let uniqs: HashSet<&Uuid> = cache_usage.iter().collect();
```
Comment (gavofyork) [inline util/src/keys/directory.rs]: isn't there a much more FP-orientated way of removing a bunch of items from `cache` based upon a predicate?

**Comment by gavofyork** [comment]
minor grumble. looks reasonable in general.

File: util/src/keys/store.rs [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+
+						let mut val = vec![0u8; key_file.crypto.cipher_text.len()];
+						match key_file.crypto.cipher_type {
+							CryptoCipherType::Aes128Ctr(ref iv) => {
+								crypto::aes::decrypt(&derived_left_bits, &iv.as_slice(), &key_file.crypto.cipher_text, &mut val);
+							}
+						}
+						val
+					}
+					_ => { unimplemented!(); }
```
Comment (arkpar) [inline util/src/keys/store.rs]: Any plans to implement soon? otherwise return an error

Comment (NikVolf) [inline util/src/keys/store.rs]: yep
will implement of importing from geth/eth formats and old versions
next part!

File: util/src/keys/directory.rs (hunk 2 above)
Comment (NikVolf) [inline util/src/keys/directory.rs]: well we want to store it anyway to avoid O(N^2) right?..

Comment (NikVolf) [inline util/src/keys/directory.rs]: improved it a bit

**Pull request merged**